de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/FileSystemChangeListener.java

```diff
@@ -1,529 +1,863 @@
 
-        SPath newSPath = new SPath(project, newPath);
-        //we handle this as a move activity
-        if (newFile.isFile()) {
-            generateFileMove(oldSPath, newSPath, false);
-        } else {
-            generateFolderMove(oldSPath, newSPath, false);
-        }
-    }
+            activity = new FileActivity(user, Type.MOVED,
+                FileActivity.Purpose.ACTIVITY, newFilePath, oldFilePath, null,
+                encoding);
 
-    @Override
-    public void fileCopied(
-        @NotNull
-        VirtualFileCopyEvent virtualFileCopyEvent) {
-        if (!enabled) {
-            return;
-        }
+            editorManager.replaceAllEditorsForPath(oldFilePath, newFilePath);
 
-        VirtualFile virtualFile = virtualFileCopyEvent.getFile();
-        File newFile = new File(virtualFile.getPath());
+            annotationManager.updateAnnotationPath(oldFilePath.getFile(),
+                newFilePath.getFile());
 
-        if (incomingFilesToFilterFor.remove(newFile)) {
-            return;
-        }
+        } else if (newPathIsShared) {
+            //moved file into shared module
+            byte[] fileContent = getContent(oldFile);
+
+            SPath newFilePath = new SPath(newParentPath.getProject(),
+                newParentPath.getProjectRelativePath().append(relativePath));
+
+            activity = new FileActivity(user, Type.CREATED,
+                FileActivity.Purpose.ACTIVITY, newFilePath, null, fileContent,
+                encoding);
+
+            if (fileIsOpen) {
+                Editor editor = projectAPI.openEditor(oldFile, false);
+
+                editorManager.addEditorMapping(newFilePath, editor);
+            }
+
+        } else if (oldPathIsShared) {
+            //moved file out of shared module
+            activity = new FileActivity(user, Type.REMOVED,
+                FileActivity.Purpose.ACTIVITY, oldFilePath, null, null, null);
+
+            editorManager.removeAllEditorsForPath(oldFilePath);
 
-        IPath path = IntelliJPathImpl.fromString(newFile.getPath());
-        IntelliJProjectImpl project = getProjectForResource(path);
+            annotationManager.removeAnnotations(oldFilePath.getFile());
+
+        } else {
+            //neither source nor destination are shared
+            if (LOG.isTraceEnabled()) {
+                LOG.trace(
+                    "Ignoring non-shared resource move - resource: " + oldFile
+                        + ", old base parent: " + oldBaseParent
+                        + ", new base parent: " + newBaseParent);
+            }
 
-        if (project == null || !isValidProject(project) ||
-            !isCompletelyShared(project)) {
             return;
         }
 
-        path = makeAbsolutePathProjectRelative(path, project);
+        fireActivity(activity);
 
-        SPath spath = new SPath(project, path);
+        if (oldPathIsShared) {
+            if (fileIsOpen) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Merge this if statement with the enclosing one. ([squid:S1066](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1066))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=307&issues=AWcYB0JPkEeD8QRk19vP&open=AWcYB0JPkEeD8QRk19vP)

[ISSUE_KEY]: AWcYB0JPkEeD8QRk19vP